[scitedotai/scite#4290] Expose tally and sectiontally in badge

u-ashish opened the pull request:
Relies on: https://github.com/scitedotai/scite-widget/pull/34

- Extend the config so that we can show the `Tally`, `SectionTally`, or _both_
- `tallyShow` is a new option to configure showing or hiding the smart citations tally. Defaults to `true` so it should work as-is for existing badges
- `sectionTallyShow` is a new option to configure showing or hiding the sections tally. Defaults to `false` so it doesn't break older badges
- `showLogo` is a new option to configure showing the logo for section tallies. Useful to hide it if we show both tallies.
- `sectionTallyLayout` is a new option to separately control the layout of the `SectionTally` component
- `chartType` is a new option (`[null, 'pie', 'bar']`) which lets you specify which chart to show for `SectionTally`
- When we retrieve the `currentDOIs`, also fetch the `sectionTallies` ( **note:** see if I can make this option to only fetch section stuff if there is a section related badge )

## Verification:

I added these to the bottom of the page

<img width="582" alt="Screen Shot 2022-04-22 at 10 31 14 AM" src="https://user-images.githubusercontent.com/14264791/164747375-0cbf217b-1623-4719-b78f-bd864d193a2f.png">


## TODO:
- [x] Maybe more efficient tally fetching so we only get section tally info if needed
- [x] The `Tally` component has a border around it, but the `SectionTally` one doesn't lol. Trying to figure out how to handle that in the different scenarios 😵 **maybe good to see Josh feedback when he is back tomorrow?**

## Testing locally:
- You need to update the `package.json` to have the branch from https://github.com/scitedotai/scite-widget/pull/34

domenicrosati commented on src/main.js:
```diff
@@ -89,8 +109,13 @@ function getDOI (el) {
+export function insertBadge (el, tally, notices, sectionTally) {
   const config = getConfig(el)
+
+  const showTally = config.showTally || false
```
We will need to make this default to true to make it backward comparable.

u-ashish replied:
Oops, it should be true, I reverted it to debug nice catch

domenicrosati commented on src/main.js:
```diff
@@ -224,14 +328,15 @@ export async function insertBadges ({ forceReload } = {}) {
     const currentDOIs = dois.slice(i * BATCH_SIZE, (i + 1) * BATCH_SIZE)
-    const [{ tallies }, { notices }] = await Promise.all([fetchTallies(currentDOIs), fetchNotices(currentDOIs)])
+
+    const [{ tallies }, { notices }, { tallies: sectionTallies }] = await Promise.all([fetchTallies(currentDOIs), fetchNotices(currentDOIs), fetchSectionTallies(currentDOIs)])
```
can we construct this depending on the config?
That way folks who only want citation type tallies and not section tallies will not have to have the extra network calls and vice versa.

domenicrosati replied:
oh didn't see `Maybe more efficient tally fetching so we only get section tally info if needed` lol

u-ashish replied:
Yep

u-ashish replied:
Okay @domenicrosati  here is what I did:

- Get the `badgesToLoad`
- For each badge, `getDOI`
- If the badge has `tallyShow`, push it into the list of DOIs to fetch tallies for
- If the badge has `sectionTallyShow`, push it into the list of DOIs to fetch section tallies for
- In the for loop where we fetch, set the `pageSize` to be the larger page size of the two lists
- Iterate in batch size, fetch the tallies, and load them
- If we exceed the smaller list's page size, `slice()` will return `[]`, and the tallies won't be fetched
- When inserting the badge, check to make sure the `doi` is in either the `tallies` or `sectionTallies` keys

I opted to do it this way so we still have one loop where it happens at the same time, and we don't have weird issues with the `fetched` attribute on the badge.

domenicrosati commented on src/test-page.js:
```diff
@@ -77,10 +77,15 @@ const Badge = ({ doi, layout, showLabels, forceCollapse, placement, small, autol
+    data-show-section-tally={showSectionTally}
+    data-chart-type={chartType}
+    data-show-logo-section-tally={showLogoSectionTally}
+    data-layout-section-tally={layoutSectionTally}
```
I will comment on these here since it's easier to see the API as the user sees them.

Some suggestions for clarity and reduction are:

**one**
Can we simply have one show logo param and controls whether the logo is show at all?
I feel like it doesn't have to be specific to the section tally and maybe is a bit confusing to a user who see's it as a option for one and not another.

**two**
I still am not sure about this but maybe the API naming scheme should be `{type_of_tally}-{param}`?
for instance :

```suggestion
    data-tally-show={showTally}
    data-section-tally-show={showSectionTally}
    data-section-tally-layout={layoutSectionTally}
```

I really don't know if it makes things clearer at all... but then maybe for the user they have a bit more of an alphabetical scheme in their head for the config?

u-ashish replied:
1. Actually yes, just thought the same, I think I did it separately since I was treating it separately but it can be composed to just 1.
2. Sure, wfm

u-ashish replied:
Updated!

u-ashish commented:
Oh also, I tested it by setting `BATCH_SIZE=1` instead of `=500` to make sure my for loop changes work.

domenicrosati reviewed (approved):
TBH looks great.

I believe the biggest proof points is lets push it to stage and test it in the extension and then we can deploy to prod if it works.
```
$ git tag v1.2.0-stage -m "My cool new version"
$ git push origin v1.2.0-stage
```

should work when you ready then in the extension just swap https://github.com/scitedotai/scite-extension/blob/fb72fcf929aecdb31f9dca745079b7f1c4945ddb/package.json#L55 for the correct branch.

I suppose we can actually keep that frozen and update another day but I think the insertBadges bulk thing is mainly used by the extension on google shcolar so it would be good to test that before wrapping up.

u-ashish commented:
@domenicrosati made a few changes since you approved:

- Added `DonutChart` to the test example
<img width="1014" alt="Screen Shot 2022-04-26 at 1 26 50 PM" src="https://user-images.githubusercontent.com/14264791/165391262-ae76ec5f-c208-4d80-85fc-52d285a8f3c9.png">


- Added section tally tooltip
<img width="525" alt="Screen Shot 2022-04-26 at 3 45 45 PM" src="https://user-images.githubusercontent.com/14264791/165391248-aebffe45-96c9-4b11-89e9-29a6e572ae05.png">

- Tweaked the tooltip footer copy

If you wanna :eyes: again
